vignettes/data.Rmd

```diff
@@ -0,0 +1,186 @@
+```
+
+As in the [`mcstate` vignette](https://mrc-ide.github.io/mcstate/articles/kalman.html) we some way of comparing model output to data.  The function used there is:
+
+```{r}
+volatility_compare <- function(state, observed, pars) {
+  dnorm(observed$observed, pars$gamma * drop(state), pars$tau, log = TRUE)
+}
+```
+
+i.e., the probability is normally distributed with mean of the equal to `gamma` multiplied by the modelled value, standard deviation of `tau` and evaluated at the observed value.  Our aim here is to adapt this so that it is implemented as part of the C++ model. This requires:
+
+* describing the data (here it will be a single numeric value, but the data could be arbitrarily complicated)
+* implementing a method to do the comparison
+* describe how to marshal the data into this structure
+
+```{r, echo = FALSE, results = "asis"}
+cc_output(readLines(system.file("examples/volatility.cpp", package = "dust")))
+```
+
+The first part, the data definition, is the definition
+
+```cc
+  struct data_t {
+    real_t observed;
+  };
+```
+
+which replaces the usual
+
+```cc
+  typedef dust::no_data data_t;
+```
+
+This structure can contain whatever you want, including things like a `std::vector` of values. In our use we've typically only had `real_t` and `int` though, even for complex models.
+
+The comparison is implemented by the method `compare_data`
+
+```cc
+  real_t compare_data(const real_t * state, const data_t& data,
+                      dust::rng_state_t<real_t>& rng_state) {
+    return dust::dnorm(data.observed, shared->gamma * state[0],
+                       shared->tau, true);
+  }
+```
+
+This looks very much like the R version above:
+
+* `dnorm` is replaced with the dust function `dust::dnorm` (do not use R API functions here, as this will be evaluated in a multi-threaded context)
+* the random number generator is available here, if your comparison is itself stochastic
+* unlike the R version where the comparison returns a vector across all particles, the C++ comparison is done for each particle
+
+Finally, the data marshalling is done by the `dust_data` template, within the `dust` namespace
+
+```cc
+namespace dust {
+template <>
+volatility::data_t dust_data<volatility>(cpp11::list data) {
+  return volatility::data_t{cpp11::as_cpp<double>(data["observed"])};
+}
+}
+```
+
+Here, you can use any function you could use from `cpp11`, much like within the `dust_pars` template specialisation. The input will be a list, corresponding to the data *at a single time point*.  The data.frame above will first be processed with `dust::dust_data`, so the first few entries look like:
+
+```{r}
+head(dust::dust_data(data), 3)
+```
+
+This is definitely a bit fiddly!  If using `odin.dust`, then this would be somewhat simplified as you could provide a single C++ file containing something like
+
+```cc
+// [[odin.dust::compare_data(observed = real_t)]]
+// [[odin.dust::compare_function]]
```

Comment:
Are these documented somewhere?

Reply:
Seems not, I'll add a small note below and have opened https://github.com/mrc-ide/odin.dust/issues/89